sql/codegen.go

```diff
@@ -225,307 +217,37 @@ func removeLastSemicolon(s string) string {
 	return s
 }
 
-func genTF(w io.Writer, pr *extendedSelect, fts fieldTypes, db *DB) error {
-	r, e := newFiller(pr, fts, db)
+func codegen(w io.Writer, pr *extendedSelect, fts fieldTypes, db *DB) error {
+	// 1. generate CommonFiller
+	filler, e := newCommonFiller(pr, fts, db)
 	if e != nil {
 		return e
 	}
-	if e = codegenTemplate.Execute(w, r); e != nil {
-		return fmt.Errorf("genTF: failed executing template: %v", e)
-	}
-	return nil
-}
-
-const codegenTemplateText = `
-import os
-# Disable Tensorflow INFO and WARNING logs
-os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
 
-import sys, json
-import tensorflow as tf
-import numpy as np
-import functools
-try:
-    import sqlflow_models
-except:
-    pass
-
-from sqlflow_submitter.db import connect, insert_values, db_generator
-
-# Disable Tensorflow INFO and WARNING
-import logging
-tf.get_logger().setLevel(logging.ERROR)
-
-` +
-	// TODO(typhoonzero): get NUM_BUCKETS, EMBEDDING_WIDTH from Extended SQL statements in
-	// COLUMN sub clause
-	`
-BATCHSIZE = 1
-EPOCHS = None
-NUM_BUCKETS=160000
-EMBEDDING_WIDTH=128
-
-train_args = dict()
-{{range $key, $value := .Attrs}}
-{{if eq $key "BATCHSIZE"}}
-BATCHSIZE = {{$value}}
-{{else if eq $key "EPOCHS"}}
-EPOCHS = {{$value}}
-{{else}}
-train_args["{{$key}}"] = {{$value}}
-{{end}}
-{{end}}
-
-driver="{{.Driver}}"
-{{if ne .Database ""}}
-database="{{.Database}}"
-{{else}}
-database=None
-{{end}}
-
-conn = connect(driver, database, user="{{.User}}", password="{{.Password}}", host="{{.Host}}", port={{.Port}})
-
-{{$iskeras := .IsKerasModel}}
-
-feature_columns = dict()
-{{ range $target, $colsCode := .FeatureColumnsCode }}
-feature_columns["{{$target}}"] = []
-{{ range $col := $colsCode }}
-feature_columns["{{$target}}"].append({{$col}})
-{{ end }}
-{{ end }}
-
-
-feature_column_names = [{{range .X}}
-"{{.FeatureName}}",
-{{end}}]
-
-
-classifier = {{.Estimator}}(
-    **feature_columns,
-    **train_args,
-    {{if .IsKerasModel}}
-)
-    {{else}}
-    model_dir = "{{.Save}}")
-    {{end}}
+	// 2. select estimatorGenerator
+	var execFn func(w io.Writer) error
+	// Try newXGBGenerator, if filler don't adapt to it, continue to try other generators below.
+	if temp, err := newXGBGenerator(filler); err != nil {
```

Comment:
We need an explicit way to distinguish the generator, the env or the estimator prefix.